### PR TITLE
[Index-Chat]: add Index model support

### PR DESCRIPTION
**Pull request by cf-icehzgzh**
本 PR 完成 `IndexTeam/Index-1.9B-Chat` 从 Hugging Face Transformers 到 PaddleFormers 的迁移，包含模型组网、HF checkpoint 自动转换、Tokenizer、前向精度、KV Cache、生成、CausalLM loss 和单卡 LoRA 训练验证。

## 1. 模型实现
- 模型：`IndexTeam/Index-1.9B-Chat`
- 架构：Decoder-only CausalLM，采用 RoPE、RMSNorm、SwiGLU MLP 和 Multi-Head Attention。
- 配置：
  - `hidden_size: 2048`
  - `intermediate_size: 5888`
  - `num_hidden_layers: 36`
  - `num_attention_heads: 16`
  - `num_key_value_heads: 16`
  - `vocab_size: 65029`

- 支持：
    - `IndexForCausalLM`；
    - CausalLM 前向；
    - KV Cache；
    - greedy generation；
    - shifted causal loss、`-100` ignore label 和 `loss_mask`；
    - LoRA 单卡训练、adapter 保存/加载和 merge。

- 模型输出头支持官方 `NormHead` 语义：
    - 训练阶段按最后一维归一化 lm head 权重；
    - eval 首次前向将归一化权重写回；
    - 官方 checkpoint 使用 `norm_head=true`，lm head 权重布局为 `[vocab_size, hidden_size]`。

## 2. 标准接口支持
- `AutoConfig.from_pretrained()` 可识别原始 HF checkpoint 的：
    - `model_type: "index"`
    - 并解析为 `IndexConfig`。
- `AutoModel.from_pretrained()` 可解析为 `IndexModel`。
- `AutoModelForCausalLM.from_pretrained()` 可解析为 `IndexForCausalLM`。
- `AutoTokenizer.from_pretrained()` 可加载 `IndexTokenizer`。
- Tokenizer 基于官方 SentencePiece 实现：
    - 不自动添加 BOS / EOS；
    - 保留 `reserved_0 = 3`、`reserved_1 = 4`；
    - 已使用官方 `tokenizer.model` 完成真实加载验证。
    - 
## 3. 权重转换
使用官方 `Index-1.9B-Chat` PyTorch checkpoint 完成转换和加载验证。
- Attention 和 MLP 线性层按 HF / Paddle 权重布局完成转置：
    - `q_proj`
    - `k_proj`
    - `v_proj`
    - `o_proj`
    - `gate_proj`
    - `up_proj`
    - `down_proj`
- `NormHead` 的 `lm_head.weight` 保持官方 `[vocab_size, hidden_size]` 布局，不进行转置。
- 转换后的模型成功加载，参数量：
```text
2,172,819,456
```

- 本 PR 不包含转换后的权重文件。

## 4. 前向精度对齐
使用相同真实 checkpoint、相同 FP32 输入，对比官方 PyTorch Transformers 4.39.2 实现与 PaddleFormers logits。

| 场景 | 指标 | 结果 |
|---|---|---|
| 完整前向 | max abs diff | `2.2888184e-05` |
| 完整前向 | mean abs diff | `2.3325802e-06` |
| 完整前向 | token argmax | 完全一致 |
| KV Cache 前向 | max abs diff | `9.417534e-06` |
| KV Cache 前向 | mean abs diff | `1.8813768e-06` |
| KV Cache 前向 | token argmax | 完全一致 |
| Paddle cache / 直接前向 | max abs diff | `9.775162e-06` |
两端 logits 均为有限值，完整前向和 KV Cache 路径均通过对齐验证。

## 5. 模型生成对齐
使用相同真实 checkpoint、相同 prompt 和 greedy decoding，对比新增 token。
PyTorch 与 Paddle 连续生成的 16 个 token 完全一致：
```text
[259, 54098, 45400, 63601, 54098, 45400, 63601, 758,
 16902, 2997, 619, 3924, 63942, 26551, 4937, 1853]
```
生成文本：
```text
PaddlePaddle是一个开源深度学习框架，由百度研发
```
结论：`Match: True`。

## 6. CausalLM loss 对齐
使用同一真实输入、相同 labels 和 `-100` ignore label，对比 PyTorch 和 Paddle 的 FP32 CausalLM loss。

| 指标 | 结果 |
|---|---|
| PyTorch loss | `3.2615509033203125` |
| Paddle loss | `3.261552095413208` |
| absolute diff | `1.1920928955078125e-06` |
| relative diff | `3.6549878595923254e-07` |
结论：两端 logits 均为有限值，loss 数值对齐通过。

## 7. 单卡 LoRA 训练验证
使用真实 `Index-1.9B-Chat` 权重，在 GPU 1 上完成 BF16 LoRA 单步训练验证。
LoRA target modules：
```text
q_proj
k_proj
v_proj
o_proj
gate_proj
up_proj
down_proj
```

验证结果：
```text
trainable_tensor_count: 504
trainable_parameter_count: 11575296
all_trainable_are_lora: True
loss_before: 13.551523208618164
updated_lora_tensor_count: 504
max_lora_delta: 0.00010000001202570274
logits_finite: True
```

同时完成：
- adapter 保存与重载验证：
    - reload logits：`max_abs = 0`
    - reload token argmax：完全一致。
- BF16 LoRA merge 验证：
    - 修复 base weight 为 BF16、LoRA delta 为 FP32 时 merge/unmerge 的 dtype 写回问题；
    - adapter 分支与 merge 分支生成 token argmax 一致。
说明：BF16 原地 merge 会产生逐层舍入误差，因此本 PR 不声明 adapter 分支与 merge 分支 logits 严格数值一致。

## 8. 单元测试与风格检查

新增测试覆盖：
- `IndexConfig`；
- `AutoConfig`、`AutoModel`、`AutoModelForCausalLM`；
- logits 和 embedding；
- shifted causal loss；
- `-100` ignore label；
- `loss_mask`；
- 空监督 loss；
- 2D padding mask 和 4D causal mask；
- KV Cache；
- `return_dict=False` tuple 返回；
- `NormHead`；
- lm head 权重转换规则；
- save/load；
- generation；
- 真实 tokenizer；
- BF16 LoRA merge/unmerge 回归。

执行：
```shell
CUDA_VISIBLE_DEVICES=1 \
INDEX_TOKENIZER_DIR=/home/user/cv/hzg/Index-1.9B-Chat \
python -m pytest \
  -p no:cacheprovider \
  tests/transformers/index/test_modeling.py \
  tests/transformers/index/test_tokenizer.py \
  tests/peft/test_lora.py::TestLoraLayer::test_merge_unmerge_bf16_weight_with_fp32_lora \
  -q
```

结果：
```text
9 passed
```
同时：
```shell
make lint
make lint
```
两遍均通过，`git diff --check` 已通过。

## 9. 提交范围

本 PR 仅包含：
- Index 模型、配置和 Tokenizer；
- AutoConfig、AutoModel 和 AutoTokenizer 必要注册；
- Index LoRA target modules；
- BF16 LoRA merge/unmerge 必要修复及回归测试；
- Index 模型和 Tokenizer 单测。

### CODE REVIEW

File: paddleformers/transformers/index/modeling.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+    def __init__(self, config: IndexConfig):
+        super().__init__(config)
+        self.config = config
+        self.model = IndexModel(config)
+        self.vocab_size = config.vocab_size
+        self.lm_head = (
+            NormHead(config.hidden_size, config.vocab_size)
+            if config.norm_head
+            else nn.Linear(config.hidden_size, config.vocab_size, bias_attr=False)
```
Comment (risemeup1111) [inline paddleformers/transformers/index/modeling.py]: P1 **优先级：P1**

**处理要求：请针对该评论修复并提交新的 commit。**

这里用 `nn.Linear(config.hidden_size, config.vocab_size)` 会让 `tie_word_embeddings=True` 的配置变成不可用：Paddle `nn.Linear` 的权重布局是 `[hidden_size, vocab_size]`，但 `tie_weights()` 会把它直接替换成 `embed_tokens.weight`，其布局是 `[vocab_size, hidden_size]`。这样 tied 配置前向时会拿错误形状的权重做 linear，导致 shape mismatch 或输出维度错误；当前测试只断言权重对象相同，没有覆盖 tied 配置的一次 forward。

请把非 `NormHead` 分支改成和 embedding 一样的 `[vocab_size, hidden_size]` 布局（例如使用 `GeneralLMHead` 或等价的 matmul head），并同步调整非 norm 的 AOA 映射和 tied forward 测试。实现形状可以类似：

```python
class IndexLMHead(nn.Layer):
    def __init__(self, hidden_size, vocab_size):
        super().__init__()
        self.weight = self.create_parameter(shape=[vocab_size, hidden_size], ...)

    def forward(self, hidden_states):
        return paddle.matmul(hidden_states, self.weight, transpose_y=True)
```

如果不打算支持 `tie_word_embeddings=True`，也需要在模型初始化时显式报错，并移除/调整当前 tied 语义测试，避免保存出一个看似支持但前向不可用的配置。

File: paddleformers/transformers/index/modeling.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+        inputs_embeds: paddle.Tensor | None = None,
+        labels: paddle.Tensor | None = None,
+        loss_mask: paddle.Tensor | None = None,
+        use_cache: bool | None = None,
+        past_key_values: Cache | None = None,
+        output_hidden_states: bool | None = None,
+        return_dict: bool | None = None,
+        **kwargs,
+    ):
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
```
Comment (risemeup1111) [inline paddleformers/transformers/index/modeling.py:183]: P1 **优先级：P1**

**处理要求：请针对该评论修复并提交新的 commit。**

`attn_mask_start_row_indices` 这个兼容别名现在会进入 `**kwargs` 后被直接丢弃，而训练链路里仍会产出这个 key（例如 DPO trainer 会把 batch 里的 `attn_mask_start_row_indices` 原样传给模型；Llama/MiniCPM/Ernie 等 wrapper 也都在 forward 入口把它转成 `attn_mask_startend_row_indices`）。Index 在 packed/flashmask/DPO 场景下如果只收到这个别名，就会丢掉稀疏行索引，退化为未分段的 causal mask，可能让同一个 packed batch 里的样本互相可见。

请在调用 `self.model(...)` 前补齐同样的别名归一化，并加一个覆盖别名路径的测试。

```suggestion
        if kwargs.get("attn_mask_start_row_indices", None) is not None and attn_mask_startend_row_indices is None:
            attn_mask_startend_row_indices = kwargs.pop("attn_mask_start_row_indices")
        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已完成初轮审查，发现两处需要修复后再合入的问题，详情见行级评论。当前 CI 仍有任务排队/运行；我这里的本地环境缺少 Paddle/Numpy 等依赖，未能复跑新增测试，最终仍请以 CI 结果为准。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — APPROVED** [review]
已复查新提交。此前两处 P1 问题已经通过新 commit 修复：lm_head tied 布局改为 `GeneralLMHead` 并补了 forward/映射测试，`attn_mask_start_row_indices` 别名也已转发并增加覆盖测试。未发现新的阻塞问题。

当前仍有部分 CI 任务运行中；我这里未本地复跑新增测试，最终请以 CI 完成结果为准。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — APPROVED** [review]
已复查新提交。此次变更移除了 `IndexModel` 对 Llama 基类注册状态的污染，并调整了 tied AOA 映射；此前两条 P1 问题在当前 head 仍保持已修复，未发现新的阻塞问题。

当前仍有部分 CI 任务运行中；本地环境缺少 `paddle`，未能复跑新增测试，最终请以 CI 完成结果为准。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — APPROVED** [review]
已复查新提交。此次变更补齐了 `IndexModel.base_model_class = IndexModel`，使 Index base model 的 AOA 映射不再错误带上 `model.` 前缀，并新增了对应测试；此前两条 P1 问题在当前 head 仍保持已修复。未发现新的阻塞问题。

当前仍有部分 CI 任务运行中；本地环境缺少 `paddle`，未能复跑新增测试，最终请以 CI 完成结果为准。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>
